### PR TITLE
libs: add gcov framework support

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
libs: add gcov framework support

In devices without storage media, you can export data to the
command line and then generate the corresponding gcda file

It can save the result output by calling __gcov_info_to_gcda
The usage is similar to:
https://gcc.gnu.org/onlinedocs/gcc/Freestanding-Environments.html#Profiling-and-Test-Coverage-in-Freestanding-Environments

Usage:
`./tools/configure.sh qemu-armv7a:nsh`
Modify the configuration
```
+CONFIG_COVERAGE_ALL=y
+CONFIG_COVERAGE_MINI=y
+CONFIG_SYSTEM_GCOV=y
```
Run:
```
qemu-system-arm -cpu cortex-a7 -nographic -smp 4 \
     -machine virt,virtualization=off,gic-version=2 \
     -net none -chardev stdio,id=con,mux=on -serial chardev:con \
     -mon chardev=con,mode=readline -kernel ./nuttx/nuttx -semihosting -s | tee gcov.txt
./nuttx/tools/gcov_convert.py -i ./gcov.txt
./nuttx/tools/gcov.sh -t arm-none-eabi-gcov
```
## Testing
qemu

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



This PR appears to mostly meet the NuttX requirements, but is lacking in key details.  Here's a breakdown:

**Strengths:**

* **Clear Summary of the "What":** The summary explains the purpose of the change (adding gcov support), the affected area (libs), and the mechanism (exporting data to the command line and generating gcda files). The usage example and configuration instructions are helpful.
* **Provides Testing Information:**  It specifies the target (qemu) and gives basic usage instructions.  The inclusion of `gcov.txt` processing is good.

**Weaknesses:**

* **Missing Issue References:**  If this relates to any existing NuttX issues, those should be linked.
* **Incomplete Impact Assessment:** The "Impact" section is entirely empty.  This is a critical part of the PR.  At a minimum, address *all* the listed impact points (user, build, hardware, documentation, security, compatibility).  Even if the answer is "NO", state it explicitly. For example: "Impact on user: NO".  If the impact *is* "YES", *explain* the impact.
* **Insufficient Testing Detail:** While testing information is present, it's minimal.  
    * **Build Host:** Provide specifics about the build host OS, CPU architecture, and compiler version used.
    * **Target Details:** Be more specific about the qemu configuration (e.g., `qemu-system-arm -cpu cortex-a7 ...`).
    * **Missing "Before" and "After" Logs:** The code blocks for "Testing logs before change" and "Testing logs after change" are empty.  Include relevant log snippets that demonstrate the functionality before and after your changes. This is crucial for reviewers to understand the impact of the changes. Show that the gcov output is as expected.

**Recommendations for Improvement:**

1. **Fill out the Impact section completely.**  This is the most significant deficiency.  Consider these points:
    * **Impact on build:**  Does enabling gcov require any build system changes or new dependencies?
    * **Impact on hardware:** Is this specific to certain architectures?
    * **Impact on documentation:** Does this require documentation updates? If so, were those updates included in the PR?
    * **Impact on security:**  Does enabling code coverage introduce any security considerations?  (Probably not, but it's good to explicitly state this).
    * **Impact on compatibility:** Does this change break any existing functionality?


2. **Provide detailed build host information.** (OS, CPU, Compiler version).

3. **Include relevant "before" and "after" test logs.** Demonstrate that gcov is *not* working before the change and *is* working after the change.  Show some example gcov output.

4. **Link any related NuttX issues.**

By addressing these points, the PR will be much more complete and easier for reviewers to evaluate.

**Review by acassis — APPROVED** [review]
Very nice @Gary-Hobson your commit log message included all needed information to let people start to using it.

File: libs/libbuiltin/libgcc/Make.defs [hunk 1]
```diff
@@ -19,10 +19,12 @@
 ############################################################################
 
 ifeq ($(CONFIG_PROFILE_MINI),y)
-
 CSRCS += profile.c
+endif
+
+ifeq ($(CONFIG_COVERAGE_MINI)$(CONFIG_ARCH_TOOLCHAIN_GCC),yy)
+CSRCS += gcov.c
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/Make.defs:26]: coverage.c

File: libs/libbuiltin/libgcc/CMakeLists.txt [hunk 2]
```diff
@@ -22,3 +22,8 @@ if(CONFIG_PROFILE_MINI)
   nuttx_add_system_library(libgprof)
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/CMakeLists.txt:22]: libgprof->libprofile

**Review by cederom — APPROVED** [review]
Thank you @Gary-Hobson :-)

File: libs/libbuiltin/libgcc/gcov.c [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+{
+  FAR const char *name = info->filename;
+  FAR uint8_t *buffer;
+  size_t size;
+  size_t osize;
+
+  filename_fn(name, arg);
+  size = gcov_buff_size(info);
+
+  buffer = malloc(size);
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/gcov.c:274]: lib_malloc

File: libs/libbuiltin/libgcc/gcov.c [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  if (!buffer)
+    {
+      syslog(LOG_ERR, "gcov alloc failed!");
+      return;
+    }
+
+  osize = gcov_convert(buffer, info);
+  if (size != osize)
+    {
+      syslog(LOG_ERR, "gcov read data failed!");
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/gcov.c:284]: lib_free

File: libs/libbuiltin/libgcc/gcov.c [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+      return;
+    }
+
+  dump_fn(buffer, size, arg);
+  free(buffer);
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/gcov.c:289]: lib_free

File: tools/gcov_convert.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+############################################################################
+
+import argparse
+import re
+
+
```
Comment (xiaoxiang781216) [inline tools/gcov_convert.py:25]: let's add a comment to describe the file format

File: libs/libbuiltin/libgcc/gcov.c [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+{
+}
+
+void __gcov_dump(void)
+{
+}
+
+void __gcov_reset(void)
+{
+  FAR struct gcov_info *info = NULL;
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:221]:
```suggestion
  FAR struct gcov_info *info;
```

File: libs/libbuiltin/libgcc/gcov.c [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+  uint32_t values;
+
+  for (info = __gcov_info_start; info; info = info->next)
+    {
+      for (functions = 0; functions < info->n_functions; functions++)
+        {
+          gfi = info->functions[functions];
+          gci = gfi->ctrs;
+
+          for (counts = 0; counts < GCOV_COUNTERS; counts++)
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:235]:
```suggestion
```

File: libs/libbuiltin/libgcc/gcov.c [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+            {
+              for (values = 0; values < gci->num; values++)
+                {
+                  gci->values[values] = 0;
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:239]: memset

File: libs/libbuiltin/libgcc/gcov.c [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+  struct gcov_ctr_info ctrs[1];    /* Instrumented counters */
+};
+
+/****************************************************************************
+ * Private Functions Prototypes
+ ****************************************************************************/
+
+static size_t gcov_convert(FAR uint8_t *buffer,
+                           FAR const struct gcov_info *info);
+static size_t gcov_buff_size(FAR const struct gcov_info *info);
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:83]:
```suggestion

```

File: libs/libbuiltin/libgcc/gcov.c [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+static size_t gcov_buff_size(FAR const struct gcov_info *info)
+{
+  uint32_t iter;
+  uint32_t iter_1;
+  uint32_t iter_2;
+
+#ifdef GCOV_12_FORMAT
+  uint32_t size = sizeof(uint32_t) * 4;
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:115]: why 4 or 3? please add some comments or define to macro

File: libs/libbuiltin/libgcc/gcov.c [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+#else
+  uint32_t size = sizeof(uint32_t) * 3;
+#endif
+
+  for (iter = 0; iter < info->n_functions; iter++)
+    {
+      size += (sizeof(uint32_t) * 5);
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:122]: comments or macro for why 5?

**Review by anchao — CHANGES_REQUESTED** [review]

File: libs/libbuiltin/libgcc/gcov.c (hunk 10 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c]: done

File: libs/libbuiltin/libgcc/gcov.c (hunk 12 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c]: done

File: libs/libbuiltin/libgcc/gcov.c (hunk 11 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c]: This is the calculation buffer size according to the gcda file format.

The gcov_buff_size function has been deleted. We can get the buffer size through the gcov_convert function.
The meaning and size of each field is easy to see in the gcov_convert function

File: libs/libbuiltin/libgcc/Make.defs (hunk 1 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/Make.defs:26]: This cannot have the same name as coverage.c in compiler-rt. When compiling using makefile, it will search for coverage.c in the compiler-rt directory, causing it to not find the correct symbol when linking.

File: libs/libbuiltin/libgcc/CMakeLists.txt (hunk 2 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/CMakeLists.txt]: done

File: libs/libbuiltin/libgcc/gcov.c (hunk 8 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c:215]: The code has been modified. The for loop in the original code is not effective. 
The latest code requires a for loop to traverse gcov_ctr_info and clear the data.

Reference Links
https://github.com/torvalds/linux/blob/master/kernel/gcov/gcc_4_7.c#L218

File: libs/libbuiltin/libgcc/Make.defs (hunk 1 above)
Comment (xiaoxiang781216) [reply libs/libbuiltin/libgcc/Make.defs:26]: let's use the full path

File: libs/libbuiltin/libgcc/gcov.c [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+void __gcov_merge_add(FAR gcov_type *counters, unsigned int n_counters)
+{
+}
+
+void __gcov_exit(void)
+{
+}
+
+void __gcov_dump(void)
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/gcov.c:198]: why not implement __gcov_dump

**Comment by acassis** [comment]
@Gary-Hobson please consider creating later a Guide (https://nuttx.apache.org/docs/latest/guides/index.html) to use GCOV with NuttX using these commands you included in the commit log.

**Comment by Gary-Hobson** [comment]
> @Gary-Hobson please consider creating later a Guide (https://nuttx.apache.org/docs/latest/guides/index.html) to use GCOV with NuttX using these commands you included in the commit log.

Yes, we will upload the documentation for checking code coverage in different environments later

Currently, we support gcc, clang, simulator, device, it will take some time to verify all environments

File: libs/libbuiltin/libgcc/gcov.c (hunk 13 above)
Comment (xiaoxiang781216) [reply libs/libbuiltin/libgcc/gcov.c:198]: but this is mini version, how can we enable toolchain and mini version at the same time?

File: libs/libbuiltin/libgcc/gcov.c [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void dump_counter(FAR void *buffer, FAR size_t *off, uint64_t v)
+{
+  if (buffer)
+    {
+      memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:105]:
```suggestion
      *(FAR uint64_t *)((FAR uint8_t *)buffer + *off) = v;
```

File: libs/libbuiltin/libgcc/gcov.c [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+  *off += sizeof(uint64_t);
+}
+
+static void dump_unsigned(FAR void *buffer, FAR size_t *off, uint32_t v)
+{
+  if (buffer)
+    {
+      memcpy((FAR uint8_t *)buffer + *off, (FAR uint8_t *)&v, sizeof(v));
```
Comment (anchao) [inline libs/libbuiltin/libgcc/gcov.c:115]:
```suggestion
      *(FAR uint32_t *)((FAR uint8_t *)buffer + *off) = v;
```

File: libs/libbuiltin/libgcc/gcov.c (hunk 14 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c]: Done

File: libs/libbuiltin/libgcc/gcov.c (hunk 13 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c:313]: Done

File: libs/libbuiltin/libgcc/gcov.c (hunk 15 above)
Comment (Gary-Hobson) [reply libs/libbuiltin/libgcc/gcov.c]: Done

**Review by anchao — APPROVED** [review]
